libtrip-notification/src/test/java/com/mapbox/navigation/trip/notification/MapboxTripNotificationTest.kt

```diff
@@ -124,4 +136,145 @@ class MapboxTripNotificationTest {
+        verify(exactly = 2) {
+            anyConstructed<RemoteViews>().setTextViewText(any(), primaryTextLambda())
+        }
+    }
+
+    @Test
+    fun whenGoThroughStartUpdateStopCycleThenNotificationCacheDropped() {
+        val routeProgress = mockk<RouteProgress>(relaxed = true)
+        val primaryText = { "Primary Text" }
+        val bannerText = mockBannerText(routeProgress, primaryText)
+        mockUpdateNotificationAndroidInteractions()
+
+        notification.onTripSessionStarted()
+        notification.updateNotification(routeProgress)
+
+        verify(exactly = 1) { bannerText.text() }
+        verify(exactly = 2) { anyConstructed<RemoteViews>().setTextViewText(any(), primaryText()) }
+
+        notification.updateNotification(routeProgress)
+
+        verify(exactly = 2) { bannerText.text() }
+        verify(exactly = 2) { anyConstructed<RemoteViews>().setTextViewText(any(), primaryText()) }
+
+        notification.onTripSessionStopped()
+        notification.onTripSessionStarted()
+
+        verify(exactly = 2) { bannerText.text() }
+        verify(exactly = 2) { anyConstructed<RemoteViews>().setTextViewText(any(), primaryText()) }
+
+        notification.updateNotification(routeProgress)
+
+        verify(exactly = 3) { bannerText.text() }
+        verify(exactly = 4) { anyConstructed<RemoteViews>().setTextViewText(any(), primaryText()) }
+    }
+
+    @Test
+    fun whenGoThroughStartUpdateStopCycleThenStartStopSessionDontAffectRemoveViews() {
+        val routeProgress = mockk<RouteProgress>(relaxed = true)
+        val primaryText = { "Primary Text" }
+        val bannerText = mockBannerText(routeProgress, primaryText)
+        mockUpdateNotificationAndroidInteractions()
+
+        notification.onTripSessionStarted()
+        notification.updateNotification(routeProgress)
+
+        verify(exactly = 1) { bannerText.text() }
+        verify(exactly = 2) { anyConstructed<RemoteViews>().setTextViewText(any(), primaryText()) }
+
+        notification.onTripSessionStopped()
+        notification.onTripSessionStarted()
+
+        verify(exactly = 1) { bannerText.text() }
+        verify(exactly = 2) { anyConstructed<RemoteViews>().setTextViewText(any(), primaryText()) }
+
+        notification.onTripSessionStopped()
+
+        verify(exactly = 1) { bannerText.text() }
+        verify(exactly = 2) { anyConstructed<RemoteViews>().setTextViewText(any(), primaryText()) }
+
+        notification.onTripSessionStarted()
+
+        verify(exactly = 1) { bannerText.text() }
+        verify(exactly = 2) { anyConstructed<RemoteViews>().setTextViewText(any(), primaryText()) }
+    }
+
+    private fun mockUpdateNotificationAndroidInteractions() {
+        every { anyConstructed<RemoteViews>().setImageViewResource(any(), any()) } just Runs
+        every { anyConstructed<RemoteViews>().setTextViewText(any(), any()) } just Runs
+        mockkStatic(TextUtils::class)
+        every { TextUtils.isEmpty(any()) } returns false
```

Comment:
Instead of completely mocking this utility, you can mock the manuever (and its text) and provide it from the route progress. This would make the test that little bit more comprehensive.

Reply:
We will need robolectric here in that case if I'm not wrong. A reason why the whole utility mocked - interaction with android.jar on JVM throws an exception

Reply:
I didn't notice the util is from the Android SDK. In this case mocking, but running the equal logic in the mocked method would be the best imo:
```
public static boolean isEmpty(@Nullable CharSequence str) {
    return str == null || str.length() == 0;
}
```

This is going to cut down the robolectric init time which wouldn't be worth it just for this check.

Reply:
It still is going to be not the same because mockk can't capture nullable types but I change it to `isEmpty()` for sure